bin/liquidator/src/tx_sender.rs

```diff
@@ -0,0 +1,237 @@
+        } else {
+            tokio::select!(
+                _ = liq_receiver.recv() => {},
+                _ = tcs_receiver.recv() => {},
+            )
+        };
+
+        // a task must be available to process
+        // find it in global shared state, and mark it as processing
+        let task = worker_pull_task(&shared_state, id, tcs_capable_workers, only_liquidation)
+            .expect("Worker woke up but has nothing to do");
+
+        // execute the task
+        let need_rebalancing = match &task {
+            WorkerTask::Liquidation(l) => worker_execute_liquidation(&mut liquidation, *l).await,
+            WorkerTask::Tcs(t) => worker_execute_tcs(&mut tcs, t.clone()).await,
+            WorkerTask::GiveUpTcs => worker_give_up_tcs(&tcs_sender).await,
+        };
+
+        if need_rebalancing {
+            rebalance_trigger_sender.send_unless_full(()).unwrap();
+        }
+
+        // remove from shared state
+        worker_finalize_task(&shared_state, id, task, need_rebalancing);
+    }
+}
+
+async fn worker_give_up_tcs(sender: &Sender<()>) -> bool {
+    sender.send(()).await.expect("sending task failed");
+    false
+}
+
+async fn worker_execute_tcs(tcs: &mut Box<TcsState>, candidates: Vec<(Pubkey, u64, u64)>) -> bool {
+    tcs.maybe_take_token_conditional_swap(candidates)
+        .await
+        .unwrap_or(false)
+}
+
+async fn worker_execute_liquidation(
+    liquidation: &mut Box<LiquidationState>,
+    candidate: Pubkey,
+) -> bool {
+    liquidation
+        .maybe_liquidate_and_log_error(&candidate)
+        .await
+        .unwrap_or(false)
+}
+
+fn worker_pull_task(
+    shared_state: &Arc<RwLock<SharedState>>,
+    id: u64,
+    tcs_capable_workers: usize,
+    only_liquidation: bool,
+) -> anyhow::Result<WorkerTask> {
+    let mut writer = shared_state.write().unwrap();
+
+    // print out list of all task for debugging
+    for x in &writer.liquidation_candidates_accounts {
+        if !writer.processing_liquidation.contains(x) {
+            trace!(" - LIQ {:?}", x);
+        }
+    }
+
+    // next liq task to execute
+    if let Some(liq_candidate) = writer
+        .liquidation_candidates_accounts
+        .iter()
+        .find(|x| !writer.processing_liquidation.contains(x))
+        .copied()
+    {
+        debug!("worker #{} got a liq candidate -> {}", id, liq_candidate);
+        writer.processing_liquidation.insert(liq_candidate);
+        return Ok(WorkerTask::Liquidation(liq_candidate));
+    }
+
+    let tcs_todo = writer.interesting_tcs.len() - writer.processing_tcs.len();
+
+    if only_liquidation {
+        debug!("worker #{} giving up TCS (todo count: {})", id, tcs_todo);
+        return Ok(WorkerTask::GiveUpTcs);
+    }
+
+    for x in &writer.interesting_tcs {
+        if !writer.processing_tcs.contains(x) {
+            trace!("  - TCS {:?}", x);
+        }
+    }
+
+    // next tcs task to execute
+    let max_tcs_batch_size = max(1, tcs_todo / tcs_capable_workers + 1);
```

Comment:
the max() looks unneeded; and for now I think it's ok to allow the job to take all the tcs candidates, or maybe up to a fairly high static limit (20?). 

Right now, if there's 4 candidates and 4 tcs_capable_workers, it'd be split into a single candidate per worker. That makes sense generally, but avoids the jup-caching and token-modification-tracking that is the main reason the tcs batching survives this PR. 

Once those two things are resolved it'd be a large simplicity gain to drop the tcs batching.